armory/art_experimental/attacks/snr_pgd.py

```diff
@@ -1,11 +1,235 @@
+                "batch_size": 1,
+                "eps_min": 0,
+                "eps_max": 64,
+                "eps_step": 0.5,
+                "max_iter": 10,
+                "norm": "snr_db",
+                "num_random_init": 2,
+                "targeted": false,
+                "tolerance": 1,
+                "verbose": false
+            },
+            "module": "armory.art_experimental.attacks.snr_pgd",
+            "name": "SNR_PGDRange2",
+            "use_label": true
+        }
+    """
+
+    def __init__(
+        self, estimator, attack="l2", eps_min=10, eps_max=50, tolerance=1, **kwargs
+    ):
+        if eps_min > eps_max:
+            raise ValueError(f"eps_min {eps_min} > eps_max {eps_max}")
+        if "eps" in kwargs:
+            raise ValueError("Use 'eps_min' and 'eps_max' instead of 'eps'")
+        if eps_min == -np.inf or eps_max == np.inf:
+            # needed due to bisection approach
+            raise ValueError(f"eps_min {eps_min} and eps_max {eps_max} must be finite")
+
+        if attack == "l2":
+            self.Attack = SNR_PGD
+        elif attack == "linf":
+            self.Attack = SNR_PGD_Linf
+        else:
+            raise ValueError(f'attack {attack} not in ("l2", "linf")')
+        self.min_attack = self.Attack(estimator, eps=eps_min, **kwargs)
+        self.max_attack = self.Attack(estimator, eps=eps_max, **kwargs)
+        self.eps_min = eps_min
+        self.eps_max = eps_max
+        self.estimator = estimator
+        self.kwargs = kwargs
+        tolerance = float(tolerance)
+        if not (tolerance >= 0):
+            raise ValueError(f"tolerance {tolerance} must be a positive float")
+        if tolerance == 0:
+            logger.warning("Using minimum float tolerance instead of 0")
+
+        self.tolerance = tolerance
+
+    def generate(self, x, y=None, **kwargs):
+        if y is None:
+            raise ValueError("This attack requires given labels")
+        if self.estimator.predict(x).argmax() != y:
+            logger.info("Original prediction failed. Returning original x.")
+            return x
+        if self.eps_min == self.eps_max:
+            return self.min_attack.generate(x, y, **kwargs)
+
+        # test endpoints
+        x_adv = self.max_attack.generate(x, y, **kwargs)
+        if self.estimator.predict(x_adv).argmax() != y:
+            logger.info(f"Success at upper boundary eps = {self.eps_max}")
+            return x_adv
+        else:
+            logger.info(f"Failure at upper boundary eps = {self.eps_max}")
+            upper_eps = self.eps_max
+
+        x_adv = self.min_attack.generate(x, y, **kwargs)
+        if self.estimator.predict(x_adv).argmax() != y:
+            logger.info(f"Success at lower boundary eps = {self.eps_min}")
+            lower_eps = self.eps_min
+            x_best = x_adv
+            eps_best = lower_eps
+        else:
+            logger.info(f"Failure at lower boundary eps = {self.eps_min}")
+            return x_adv
```

Comment:
Wondering if we want to return `x` instead of `x_adv` to match the behavior of `SNR_PGDRange`